python/pyspark/sql/types.py

```diff
@@ -1119,8 +1120,8 @@ def Dict(d):
     class Row(tuple):
 
         """ Row in DataFrame """
-        __DATATYPE__ = dataType
-        __FIELDS__ = tuple(f.name for f in dataType.fields)
+        __datatype = dataType
```

Comment:
_restore_object is only used for Row, so the dataType should be StructType. The reason it will crash while Row has a reference to dataType is that there will be multiple id for a Row class in `_cached_cls`, all of them are the same StructType, but from different batches (see another comment).

So we should not add `__dataType__` for other datatypes